Stats: Update PWYW flow with new slider

a8ck3n opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #84859

## Proposed Changes

- Adds updated tier slider to PWYW flow.
- Refactors `TierUpgradeSlider` to be reusable across flows.
- Adds `StatsCommercialUpgradeSlider` and `StatsPWYWUpgradeSlider` for flow-specific logic.
- Updates Commercial flow to use new component.

<img width="598" alt="SCR-20231207-nwyj" src="https://github.com/Automattic/wp-calypso/assets/40267301/f6323c91-9067-4558-bea7-86fb4c79e588">

## Follow-on Changes

To come in follow-up PR(s):

- Remove the old `TierUpgradeSlider` component/file.
- Remove the old slider code from the PWYW flow. (it's hidden for now)
- Move the new `TierUpgradeSlider` to the component library.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

Currently behind the `stats/tier-upgrade-slider` flag. More to come…

1. Launch the Calypso Live link.
2. Navigate to **Stats → Traffic**.
3. Click on CTA to purchase Stats plan.
4. Add `stats/tier-upgrade-slider` to URL flags.
5. Confirm new tier slider is visible and updates the page correctly.

1. Click the "choose a commercial plan" link in the highlighted box.
2. Confirm the new tier slider shows correctly here too.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

a8ck3n committed:
Introduce new wrapper for slider.

The idea here is to pull all the context-specific logic up a level so the TierUpgradeSlider can be used across the Commercial, PWYW, and AI flows.

a8ck3n committed:
Introduce handler on purchase page.

a8ck3n committed:
Remove TierUpgradeSlider from page.

a8ck3n committed:
Push tier fetching up a level.

a8ck3n committed:
Decouple tier evaluation from slider.

a8ck3n committed:
Move UI strings up a level.

a8ck3n committed:
Further decouple steps from tiers.

a8ck3n committed:
Add additional data transform for tiers.

No real data validation in place yet. We should probably think on a strategy for that.

a8ck3n committed:
Move logic for tier messaging to parent.

a8ck3n committed:
Remove remaining tier logic from slider.

a8ck3n committed:
Move logic wrapper to new file.

The StatsCommercialUpgradeSlider serves as a logic wrapper around the TierUpgradeSlider and handles context specific to the commerical upgrade flow.

a8ck3n committed:
Move TierUpgradeSlider to new file.

This should eventually move to a components library so the AI flow can make use of it too.

a8ck3n committed:
Make the info string optional.

Not necessary for the PWYW flow.

a8ck3n committed:
Add StatsPWYWUpgradeSlider component.

Early days yet. Needs to adopt the logic from the PWYW flow for setting up the tiers.

a8ck3n committed:
Expose marks prop on slider.

a8ck3n committed:
Set marks prop from both call sites.

a8ck3n committed:
Expose an initialValue prop on the slider.

To accomodate the PWYW flow not starting at the lowest available plan.

a8ck3n committed:
Set PWYW slider to start at middle tier.

a8ck3n committed:
Set a minimum width on the slider.

This is set in the CSS for the page already but is currently not taking effect.

a8ck3n committed:
Add PWYW step logic to StatsPWYWUpgradeSlider.

Still exists in parent as we haven’t removed the previous slider implementation yet.

a8ck3n committed:
Add new tier slider to PWYW flow.

Currently behind the “stats/tier-upgrade-slider” feature flag. Should be fully hooked up though. State stays in sync between both sliders when displayed.

a8ck3n committed:
Move new slider down the page a bit.

kangzj commented:
I really like how you write your comments for commits, so organized and easy to understand 👍

a8ck3n committed:
Hide old slider for now.

a8ck3n committed:
Remove min-width value on slider.

a8ck3n committed:
Remove unused mock data.

a8ck3n committed:
Add typing for the PWYW slider settings.

Settings actually represent plan details and some UI-specific logic for emoji. Added comments on breaking this apart.

a8ck3n committed:
Add typing to steps transformation.

a8ck3n committed:
Clean up function definition a bit.

a8ck3n committed:
Translate UI strings for PWYW slider.

a8ck3n committed:
Translate recommended price.

a8ck3n committed:
Comments++

a8ck3n committed:
Update component name.

And confirm import statements point to refactored component.

a8ck3n committed:
Add typing to UI string props.

a8ck3n committed:
Add interface for tier steps.

kangzj commented on client/my-sites/stats/stats-purchase/stats-purchase-personal.tsx:
```diff
@@ -107,22 +115,34 @@ const PersonalPurchase = ( {
-					},
-				} ) }
-			</p>
+			{ showOldSlider && (
```
I feel like if it's a `isTierUpgradeSliderEnabled` and `isTierUpgradeSliderEnabled` pair, then it'd be more clear that these two paths are exclusive to one another

a8ck3n replied:
My use case was that I wanted them both to draw while I was testing to make sure I hadn't missed any behaviour. That's why I didn't make it an if/else test. The intent was to test and ship, then remove the old code paths completely.

a8ck3n replied:
I have a follow-up PR in the works to remove this and other unused code paths.

kangzj committed:
resolve conflict

kangzj committed:
cherry check style changes

kangzj committed:
fix thumb shape for wpcom

kangzj committed:
focusOnShow={ false }

kangzj commented on client/my-sites/stats/stats-purchase/stats-purchase-tier-upgrade-slider.tsx:
```diff
@@ -1,89 +1,57 @@
 	const [ currentPlanIndex, setCurrentPlanIndex ] = useState( 0 );
 	const sliderMin = 0;
-	const sliderMax = tiers?.length - 1;
+	const sliderMax = steps?.length - 1;
```
👍

kangzj commented on client/my-sites/stats/stats-purchase/stats-pwyw-uprade-slider/index.tsx:
```diff
@@ -0,0 +1,133 @@
+// TODO: Remove test data.
+// Currently used as a fallback if no plan info is provided.
+// Better approach is to require plan info and do some form of validation on it.
+function getPWYWPlanTiers( minPrice: number, stepPrice: number ) {
```
We probably want to remove the test data - I would prefer error rather than wrong data in the case. What do you think?

a8ck3n replied:
Yes, definitely. We don't want test data showing up in the shipping code. I was thinking we'd remove that once the real data was available. Removing it before makes the set up for testing that little bit harder.

Maybe a smarter approach for leaving the test data in would be to only return it if the feature-flag is enabled. That would prevent it from accidentally showing up in the shipping version.

a8ck3n replied:
I must have been confused here. The tier data isn't an issue for testing in the PWYW flow. I'll get this removed today as well.

kangzj reviewed (approved):
Awesome work!

I went merged the changes from the PRs merged earlier, like the color for the marks,  the round thumbs for wpcom sites and some tiny bits.

We probably want to remove `client/my-sites/stats/stats-purchase/stats-purchase-tier-upgrade-slider.tsx` and its CSS in follow up PR, because it's not referenced anymore.

Some nitpick left inline, and no blockers for merging!

kangzj commented:
We'll need an proper initial status for the PWYW slider

<img width="641" alt="image" src="https://github.com/Automattic/wp-calypso/assets/1425433/feef74ef-4f6f-4469-a7e3-5cec43ea0ce1">

The commercial slider is doing okay:

<img width="647" alt="image" src="https://github.com/Automattic/wp-calypso/assets/1425433/b3414412-4add-4e04-8f6f-c8e78ad6379b">

kangzj commented on client/my-sites/stats/stats-purchase/stats-pwyw-uprade-slider/index.tsx:
```diff
@@ -0,0 +1,133 @@
+			className="stats-pwyw-upgrade-slider"
+			uiStrings={ uiStrings }
+			steps={ steps }
+			initialValue={ ( steps.length - 1 ) / 2 }
```
Low priority: I think the initialValue could be exposed for flexibility.

a8ck3n replied:
I was thinking something similar. Right now the setup is handled at a few different levels but it would be nice to have it all in one place similar to how the commercial flow handles it. Noted for follow-up!

a8ck3n replied:
Added to #84859 for follow-up.

kangzj commented on client/my-sites/stats/stats-purchase/stats-purchase-tier-upgrade-slider.tsx:
```diff
@@ -97,31 +65,15 @@ function TierUpgradeSlider( {
-								},
-							}
-						) }
+					{ showPopup && popupInfoString }
```
the popup is already gated by isVisible, so I guess it wouldn't be needed to be gated again.

kangzj commented on client/my-sites/stats/stats-purchase/stats-purchase-single-item.tsx:
```diff
@@ -125,6 +125,10 @@ Thanks\n\n`;
 	// Currently displaying below the flow to maintain existing behaviour.
 	const isTierUpgradeSliderEnabled = config.isEnabled( 'stats/tier-upgrade-slider' );
 
+	const handleSliderChanged = ( value: number ) => {
```
Could we just pass `setPurchaseTierQuantity` down rather than wrapping it?

a8ck3n replied:
Yes, but there's actually a bug here that I forgot to comment on. The mock data returns view counts as numbers except for the last tier, when it's a string. I was assuming this reflects the data as provided by the live API. If that's the case, and we just forward this without special handling, the checkout fails.

I'll review this and get a working solution in place today.

Thanks for flagging this!

kangzj commented on client/my-sites/stats/stats-purchase/stats-pwyw-uprade-slider/index.tsx:
```diff
@@ -0,0 +1,133 @@
+	const marks = [ 0, steps.length - 1 ];
+
+	const handleSliderChanged = ( index: number ) => {
+		onSliderChange( index );
```
Same here

kangzj commented on client/my-sites/stats/stats-purchase/tier-upgrade-slider/index.tsx:
```diff
@@ -0,0 +1,92 @@
+				className="tier-upgrade-slider__extension-popover-wrapper"
+			>
+				<div className="tier-upgrade-slider__extension-popover-content">
+					{ showPopup && popupInfoString }
```
we probably don't need to gate it again here 😄

kangzj commented on client/my-sites/stats/stats-purchase/stats-pwyw-uprade-slider/index.tsx:
```diff
@@ -0,0 +1,133 @@
+		return String.fromCodePoint( 0x2764, 0xfe0f );
+	}
+	// Big spender! Fire emoji.
+	return String.fromCodePoint( 0x1f525 );
```
Just for your information, this never shows up - I understand it's probably because of following the new design?

a8ck3n replied:
This works correctly for me. The last 5 tiers show the fire emoji.

I wonder if the number of tiers is different based on your currency though. If that's the case, then my logic here wouldn't work as it assumes 28 total pricing tiers. The original code set the thresholds based on a percentage. I'll revisit that and see if I need to borrow from that logic.

a8ck3n replied:
Added to #84859 so I don't forget to follow up on this.

kangzj commented on client/my-sites/stats/stats-purchase/stats-pwyw-uprade-slider/index.tsx:
```diff
@@ -0,0 +1,133 @@
+	return tiers;
+}
+
+function useTranslatedStrings() {
```
I'm not saying this is bad, and I understand this would reduce noises in the component; however I personally find it handy when they are inlined - they kinda explains what the components are for and hence make it easier to understand the context.

And easier to search and locate 😄

a8ck3n replied:
Noted! That's two thumbs down for this pattern so I'll bring the strings inline. 👍

kangzj commented:
I'm merging it now. @grzegorz-cp  and @dognose24 please feel free to come and take a look, we could address any feedback in follow up PR 😄 

Again great work @a8ck3n! The pages are more like it now 🥳

a8ck3n commented:
> We'll need an proper initial status for the PWYW slider

Just noting that I'm unable to reproduce this in my testing. Will chase this up with the team in Slack.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/10069627).

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/10069627).
